File: src/main/java/org/mechaflow/Machine.java
package org.mechaflow;

import org.flowutils.time.Time;

import java.util.List;

/**
 * A machine or part of a machine.
 * Has ports that can be connected to ports on other Machines.
 */
public interface Machine {

    /**
     * @return the ports available on this Machine.
     */
    List<Port> getPorts();

    /**
     * Update the components internal state and output ports based on its internal state and input ports (and elapsed time since the last update).
     * @param time current game time and the duration since the last update.
     */
    void update(Time time);

    /**
     * Propagate updates from output ports to connected input ports.
     * @param time current game time and the duration since the last update.
     */
    void propagate(Time time);

    /**
     * Disconnects all ports.
     */
    void disconnectAll();
}

File: src/main/kotlin/org/mechaflow2/Machine.kt
package org.mechaflow2

import org.flowutils.time.Time
import org.flowutils.updating.Updating
import org.mechaflow.Port
import org.mechaflow.standard.heat.Heatable

/**
 * A machine or part of a machine.
 * Has ports that can be connected to ports on other Machines.
 */
interface Machine: Heatable, Updating {

    /**
     * @return the ports available on this Machine.
     */
    fun getPorts(): List<Port>

    /**
     * Propagate updates from output ports to connected input ports.
     * @param time current game time and the duration since the last update.
     */
    fun propagate(time: Time)

    /**
     * Disconnects all ports.
     */
    fun disconnectAll() {
        for (port in getPorts()) {
            port.disconnect()
        }
    }

}
